server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1728,7 +1732,7 @@ public static ClusterState updateWithSnapshots(ClusterState state,
                 final IndexRoutingTable indexRoutingTable = routingTable.index(indexName);
                 assert indexRoutingTable != null;
                 for (int i = 0; i < indexMetadata.getNumberOfShards(); i++) {
-                    ShardId shardId = new ShardId(indexMetadata.getIndex(), i);
+                    final ShardId shardId = indexRoutingTable.shard(i).shardId();
```

Comment:
No need to create a new shard id, especially not when we use that shard id for comparisons all over the place (doesn't help that much because we eventually read other instances back from the wire but it's a freebie IMO).